src/geop/main.py

```diff
@@ -28,26 +33,48 @@ def count():
     if not os.path.isfile(raster_path):
         return handle_error('{} is not valid file in DATA_DIR'.format(raster))
 
-    counts = {}
-    start = time.clock()
-    with rasterio.open(raster_path) as src:
-        w = src.read(1, window=((10000, 10350), (10000, 10350)))
-
-        for idx, cell in np.ndenumerate(w):
-            cell_val = str(cell)
+    query_polygon = request.get_json(silent=True)
+    if not query_polygon:
+        return handle_error('GeoJSON polygon is required in body')
 
-            if cell_val in counts:
-                counts[cell_val] += 1
-            else:
-                counts[cell_val] = 1
+    start = time.clock()
+    geom_5070 = reproject(shape(query_polygon))
 
-    elapsed = time.clock() - start
-    cells = reduce(lambda s, key: s + counts[key], counts.keys(), 0)
+    with rasterio.open(raster_path) as src:
+        # Read a chunk of the raster that contains the bounding box of the
+        # input geometry.  This has memory implications if that rectangle
+        # is large. The affine transformation maps geom coordinates to the
+        # image mask below.
+        window, shifted_affine = get_window_and_affine(geom_5070, src)
+        data = src.read(1, window=window)
+
+    # Create a numpy array to mask cells which don't intersect with the
+    # polygon. Cells that intersect will have value of 0 (unmasked), the
+    # rest are filled with 1s (masked)
+    geom_mask = features.rasterize(
+        [(geom_5070, 0)],
+        out_shape=data.shape,
+        transform=shifted_affine,
+        fill=1,
+        dtype=np.uint8,
+        all_touched=True
+    )
+
+    masked_data = np.ma.array(data=data, mask=geom_mask.astype(bool))
+
+    # Perform count using numpy built-ins.  Compressing the masked array
+    # creates a 1D array of just unmasked values.  May be able to speed up
+    # by using scipy count_tier_group, but this is working well for now
+    values, counts = np.unique(masked_data.compressed(), return_counts=True)
+
+    # Make dict of val: count with string keys for valid json
+    count_map = dict(zip(values, counts))
```

Comment:
Since it doesn't really matter what type `values` is at this stage, it might make sense to convert it before the `zip` so that the `format_counts` step below isn't needed:

```python
count_map = dict(zip(map(str, values), counts))
```